### PR TITLE
ESP32: Add DeviceInfoProviderImpl

### PR DESCRIPTION
**Pull request by wqx6**
#### Problem
Need to add DeviceInfoProviderImpl for the ESP32 platform, it is related to some test cases in TE9 (TC-LC, TC-TFT, TC-LO).

#### Change overview
Add DeviceInfoProviderImpl for ESP32 platform

#### Testing
Tested on ESP32C3, FixedLabel cluster and UserLabel cluster work well.

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/platform/ESP32/DeviceInfoProviderImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+CHIP_ERROR DeviceInfoProviderImpl::GetUserLabelLength(EndpointId endpoint, size_t & val)
+{
+    DefaultStorageKeyAllocator keyAlloc;
+    uint16_t len = static_cast<uint16_t>(sizeof(val));
+
+    return mStorage->SyncGetKeyValue(keyAlloc.UserLabelLengthKey(endpoint), &val, len);
+}
+
+CHIP_ERROR DeviceInfoProviderImpl::SetUserLabelAt(EndpointId endpoint, size_t index, const UserLabelType & userLabel)
```
Comment (bzbarsky-apple) [inline src/platform/ESP32/DeviceInfoProviderImpl.cpp:133]: So this allows unbounded storage, right?  Is that intended?  Is some sort of bound imposed higher up the callstack?

File: src/platform/ESP32/DeviceInfoProviderImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+    return true;
+}
+
+DeviceInfoProvider::SupportedLocalesIterator * DeviceInfoProviderImpl::IterateSupportedLocales()
+{
+    return new SupportedLocalesIteratorImpl();
+}
+
+size_t DeviceInfoProviderImpl::SupportedLocalesIteratorImpl::Count()
```
Comment (bzbarsky-apple) [inline src/platform/ESP32/DeviceInfoProviderImpl.cpp:214]: Again, this should be up to the application, not a "platform" service, as far as I can tell.

File: src/platform/ESP32/DeviceInfoProviderImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+}
+
+DeviceInfoProvider::SupportedCalendarTypesIterator * DeviceInfoProviderImpl::IterateSupportedCalendarTypes()
+{
+    return new SupportedCalendarTypesIteratorImpl();
+}
+
+size_t DeviceInfoProviderImpl::SupportedCalendarTypesIteratorImpl::Count()
+{
+    // In ESP32 Simulation, return the size of the hardcoded list of Strings that are valid values for the Calendar Types.
```
Comment (bzbarsky-apple) [inline src/platform/ESP32/DeviceInfoProviderImpl.cpp:291]: Again, not a "platform" service...

File: src/platform/ESP32/PlatformManagerImpl.cpp [hunk 4]
```diff
@@ -61,6 +62,7 @@ CHIP_ERROR PlatformManagerImpl::_InitChipStack(void)
 {
     SetConfigurationMgr(&ConfigurationManagerImpl::GetDefaultInstance());
     SetDiagnosticDataProvider(&DiagnosticDataProviderImpl::GetDefaultInstance());
+    SetDeviceInfoProvider(&DeviceInfoProviderImpl::GetDefaultInstance());
```
Comment (bzbarsky-apple) [inline src/platform/ESP32/PlatformManagerImpl.cpp:65]: So I guess an application could call `SetDeviceInfoProvider` itself with its own info?  But then `DeviceInfoProviderImpl` would be dead code being carted around for no reason.  The Linux/Mac impls can afford to do that carting around, maybe.  But I'm not sure we want that on esp32...

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Per comments above, this is not the right approach to a device info provider that would actually be usable for anything other than pretending to have one....

**Comment by wqx6** [comment]
> Per comments above, this is not the right approach to a device info provider that would actually be usable for anything other than pretending to have one....

This PR is only to pass the test cases.   We will update the DeviceInfoProvider like the FactoryDataProvider in #17240 and update the script to generate the binary file which has the device information (Fixed Label, SupportedCalendar, SupportedLocales). And I have added a TODO tag in this PR.

**Review by bzbarsky-apple — APPROVED** [review]
OK, if this is meant to be configurable by the application, great.

**Review by chshu — APPROVED** [review]

File: src/platform/ESP32/DeviceInfoProviderImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+
+    VerifyOrReturnError(mIndex < 4, false);
+
+    ChipLogProgress(DeviceLayer, "Get the fixed label with index:%d at endpoint:%d", mIndex, mEndpoint);
+
+    switch (mIndex)
+    {
+    case 0:
+        labelPtr = "room";
+        valuePtr = "bedroom 2";
```
Comment (andy31415) [inline src/platform/ESP32/DeviceInfoProviderImpl.cpp:80]: Why does every ESP32 device need to have these labels hardcoded?

this seems it should not be in platform

Comment (bzbarsky-apple) [reply src/platform/ESP32/DeviceInfoProviderImpl.cpp:80]: See https://github.com/project-chip/connectedhomeip/pull/18236#issuecomment-1134112446

**Comment by andy31415** [comment]
> > Per comments above, this is not the right approach to a device info provider that would actually be usable for anything other than pretending to have one....
> 
> This PR is only to pass the test cases. We will update the DeviceInfoProvider like the FactoryDataProvider in #17240 and update the script to generate the binary file which has the device information (Fixed Label, SupportedCalendar, SupportedLocales). And I have added a TODO tag in this PR.

@wqx6 - when will this change be done? A month passed already and we still have the wrong pattern.

**Comment by wqx6** [comment]
> > > Per comments above, this is not the right approach to a device info provider that would actually be usable for anything other than pretending to have one....
> > 
> > 
> > This PR is only to pass the test cases. We will update the DeviceInfoProvider like the FactoryDataProvider in #17240 and update the script to generate the binary file which has the device information (Fixed Label, SupportedCalendar, SupportedLocales). And I have added a TODO tag in this PR.
> 
> @wqx6 - when will this change be done? A month passed already and we still have the wrong pattern.

I discussed with @shubhamdp and he will add the deviceinfo to the script `connectedhomeip/scripts/tools/generate_esp32_chip_factory_bin.py` then the API changes in ESP32DeviceInfoProvider will be added.

**Pull request merged**